src/main/kotlin/com/sourcegraph/cody/edit/FixupSession.kt

```diff
@@ -0,0 +1,351 @@
+  }
+
+  private fun findRangeEnclosing(ranges: List<Range>, offset: Int): Range? {
+    return ranges.firstOrNull { range ->
+      range.start.toOffset(document) <= offset &&
+          range.end.toOffset(document) >= offset
+    }
+  }
+
+  fun update(task: EditTask) {
+    logger.warn("Task updated: $task")
+    when (task.state) {
+      CodyTaskState.Idle -> {}
+      CodyTaskState.Working,
+      CodyTaskState.Inserting,
+      CodyTaskState.Applying,
+      CodyTaskState.Formatting -> {}
+      // Tasks remain in this state until explicit accept/undo/cancel.
+      CodyTaskState.Applied -> showAcceptGroup()
+      // Then they transition to finished.
+      CodyTaskState.Finished -> {}
+      CodyTaskState.Error -> {}
+      CodyTaskState.Pending -> {}
+    }
+  }
+
+  /** Notification that the Agent has deleted the task. Clean up if we haven't yet. */
+  fun taskDeleted() {
+    finish()
+  }
+
+  private fun showLensGroup(group: LensWidgetGroup) {
+    lensGroup?.let { if (!it.isDisposed.get()) Disposer.dispose(it) }
+    lensGroup = group
+    var range = selectionRange
+    if (range == null) {
+      // Be defensive, as the protocol has been fragile with respect to selection ranges.
+      logger.warn("No selection range for session: $this")
+      // Last-ditch effort to show it somewhere other than top of file.
+      val position = Position(editor.caretModel.currentCaret.logicalPosition.line, 0)
+      range = Range(start = position, end = position)
+    } else {
+      val position = Position(range.start.line, 0)
+      range = Range(start = position, end = position)
+    }
+    group.show(range)
+    // Make sure the lens is visible.
+    ApplicationManager.getApplication().invokeLater {
+      val logicalPosition = LogicalPosition(range.start.line, range.start.character)
+      editor.scrollingModel.scrollTo(logicalPosition, ScrollType.CENTER)
+    }
+  }
+
+  private fun showWorkingGroup() {
+    showLensGroup(LensGroupFactory(this).createTaskWorkingGroup())
+  }
+
+  private fun showAcceptGroup() {
+    showLensGroup(LensGroupFactory(this).createAcceptGroup())
+  }
+
+  fun finish() {
+    try {
+      controller.removeSession(this)
+      rangeMarkers.forEach { it.dispose() }
+      rangeMarkers.clear()
+    } catch (x: Exception) {
+      logger.debug("Session cleanup error", x)
+    }
+    Disposer.dispose(this)
+  }
+
+  /** Subclass sends a fixup command to the agent, and returns the initial task. */
+  abstract fun makeEditingRequest(agent: CodyAgent): CompletableFuture<EditTask>
+
+  fun accept() {
+    CodyAgentService.withAgent(project) { agent ->
+      agent.server.commandExecute(CommandExecuteParams(COMMAND_ACCEPT, listOf(taskId!!)))
+    }
+    finish()
+  }
+
+  fun cancel() {
+    CodyAgentService.withAgent(project) { agent ->
+      agent.server.commandExecute(CommandExecuteParams(COMMAND_CANCEL, listOf(taskId!!)))
+    }
+    if (performedEdits) {
+      undo()
+    } else {
+      finish()
```

Comment:
That `finish()` can be executed before `agent.server.commandExecute(CommandExecuteParams(COMMAND_CANCEL, listOf(taskId!!)))`, as that agent code waits for agent to be available and executes on the background thread asynchronously. 
Just wanted to confirm that tis is fine?